common/src/main/java/com/microsoft/identity/common/crypto/KeyStoreBackedSecretKeyProvider.kt

```diff
@@ -0,0 +1,506 @@
+                clearKeyFromCache()
+                return null
+            }
+            return unwrapSecretKey(wrappedSecretKey, keyPair)
+        } catch (e: ClientException) {
+            // Reset KeyPair info so that new request will generate correct KeyPairs.
+            // All tokens with previous SecretKey are not possible to decrypt.
+            Logger.warn(
+                methodTag, "Error when loading key from Storage, " +
+                        "wipe all existing key data "
+            )
+            deleteSecretKeyFromStorage()
+            throw e
+        }
+    }
+
+    private fun wrapSecretKey(
+        secretKey: SecretKey,
+        keyPair: KeyPair
+    ): Pair<ByteArray, CipherSpec> {
+        val methodTag = "$TAG:wrapSecretKey"
+        val span = OTelUtility.createSpanFromParent(
+            SpanName.SecretKeyWrapping.name,
+            SpanExtension.current().spanContext
+        )
+
+        return try {
+            SpanExtension.makeCurrentSpan(span).use { _ ->
+                span.setAttribute(AttributeName.secret_key_wrapping_operation.name, "WRAP")
+                val cipherParamsSpec = getKeyPairCompatibleCipherSpecs(keyPair).firstOrNull()
+                    ?: throw ClientException(
+                        ClientException.UNKNOWN_CRYPTO_ERROR,
+                        "No compatible cipher specs found for key pair: $keyPair"
+                    )
+                span.setAttribute(
+                    AttributeName.secret_key_wrapping_transformation.name,
+                    cipherParamsSpec.transformation
+                )
+                Logger.info(methodTag, "Wrapping secret key with cipher spec: $cipherParamsSpec")
+                val wrappedKey = AndroidKeyStoreUtil.wrap(
+                    secretKey,
+                    keyPair,
+                    cipherParamsSpec.transformation,
+                    cipherParamsSpec.algorithmParameterSpec
+                )
+                span.setStatus(StatusCode.OK)
+                wrappedKey to cipherParamsSpec
+            }
+        } catch (exception: Exception) {
+            Logger.error(methodTag, "Failed to wrap secret key", exception)
+            span.setStatus(StatusCode.ERROR)
+            span.recordException(exception)
+            throw exception
+        } finally {
+            span.end()
+        }
+    }
+
+    private fun unwrapSecretKey(
+        wrappedSecretKey: WrappedSecretKey,
+        keyPair: KeyPair
+    ): SecretKey {
+        val methodTag = "$TAG:unwrapSecretKey"
+        val span = OTelUtility.createSpanFromParent(
+            SpanName.SecretKeyWrapping.name,
+            SpanExtension.current().spanContext
+        )
+
+        return try {
+            SpanExtension.makeCurrentSpan(span).use { _ ->
+                span.setAttribute(AttributeName.secret_key_wrapping_operation.name, "UNWRAP")
+                val cipherParamsSpec = getKeyPairCompatibleCipherSpecs(keyPair).firstOrNull { spec ->
```

Comment:
>getKeyPairCompatibleCipherSpecs(keyPair) [](http://example.com/codeflow?start=39&length=40)

Why do we need to first get compatible and apply second filtering based on ciphertransform?

You can directly select spec, which matches wrappedSecretKey.cipherTransformation, single filtering